### ISSUE TITLE
Cannot create PoolableConnectionFactory (FATAL: sorry, too many clients already)

### ISSUE THREAD

**Issue by fxjordan** [rwth-acis/Gamification-Framework#18]
After some requests in a short period of time (I was trying requests for 10 Minutes) there seems to be a problem with too many clients for the SQL database.

```
OIDC sub found. Authenticating...
attempting login with id: netfelix
field not found: monitor
java.sql.SQLException: Cannot create PoolableConnectionFactory (FATAL: sorry, too many clients already)
at org.apache.commons.dbcp2.BasicDataSource.createPoolableConnectionFactory(BasicDataSource.java:2152)
at org.apache.commons.dbcp2.BasicDataSource.createDataSource(BasicDataSource.java:1903)
at org.apache.commons.dbcp2.BasicDataSource.getConnection(BasicDataSource.java:1413)
at i5.las2peer.services.gamificationGameService.database.DatabaseManager.getConnection(DatabaseManager.java:31)
at i5.las2peer.services.gamificationGameService.GamificationGameService.getSeparateGameInfo(GamificationGameService.java:517)
at jdk.internal.reflect.GeneratedMethodAccessor152.invoke(Unknown Source)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.base/java.lang.reflect.Method.invoke(Method.java:564)
at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:469)
at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:391)
at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80)
at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255)
at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680)
at i5.las2peer.restMapper.RESTService.handle(RESTService.java:148)
at jdk.internal.reflect.GeneratedMethodAccessor140.invoke(Unknown Source)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.base/java.lang.reflect.Method.invoke(Method.java:564)
at i5.las2peer.execution.ServiceHelper.execute(ServiceHelper.java:139)
at i5.las2peer.security.ServiceAgentImpl.invoke(ServiceAgentImpl.java:453)
at i5.las2peer.security.ServiceAgentImpl.lambda$handle$0(ServiceAgentImpl.java:479)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: org.postgresql.util.PSQLException: FATAL: sorry, too many clients already
at org.postgresql.core.v3.ConnectionFactoryImpl.doAuthentication(ConnectionFactoryImpl.java:623)
at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:163)
at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:215)
at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:51)
at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:225)
at org.postgresql.Driver.makeConnection(Driver.java:466)
at org.postgresql.Driver.connect(Driver.java:265)
at org.apache.commons.dbcp2.DriverConnectionFactory.createConnection(DriverConnectionFactory.java:39)
at org.apache.commons.dbcp2.PoolableConnectionFactory.makeObject(PoolableConnectionFactory.java:205)
at org.apache.commons.dbcp2.BasicDataSource.validateConnectionFactory(BasicDataSource.java:2162)
at org.apache.commons.dbcp2.BasicDataSource.createPoolableConnectionFactory(BasicDataSource.java:2148)
... 35 more
Request failed: java.lang.NullPointerException
java.lang.NullPointerException
at i5.las2peer.services.gamificationGameService.GamificationGameService.getSeparateGameInfo(GamificationGameService.java:533)
at jdk.internal.reflect.GeneratedMethodAccessor152.invoke(Unknown Source)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.base/java.lang.reflect.Method.invoke(Method.java:564)
at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:469)
at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:391)
at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80)
at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255)
at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680)
at i5.las2peer.restMapper.RESTService.handle(RESTService.java:148)
at jdk.internal.reflect.GeneratedMethodAccessor140.invoke(Unknown Source)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.base/java.lang.reflect.Method.invoke(Method.java:564)
at i5.las2peer.execution.ServiceHelper.execute(ServiceHelper.java:139)
at i5.las2peer.security.ServiceAgentImpl.invoke(ServiceAgentImpl.java:453)
at i5.las2peer.security.ServiceAgentImpl.lambda$handle$0(ServiceAgentImpl.java:479)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
at java.base/java.lang.Thread.run(Thread.java:832)
```

**Comment by fxjordan** [comment]
This is still a problem and makes the cluster deployment not usable at the moment!

**Comment by fxjordan** [comment]
Latest error (line numbers have changed since last error example):
```java
OIDC sub found. Authenticating...
attempting login with id: netfelix
field not found: monitor
java.sql.SQLException: Cannot create PoolableConnectionFactory (FATAL: sorry, too many clients already)
 at org.apache.commons.dbcp2.BasicDataSource.createPoolableConnectionFactory(BasicDataSource.java:2152)
 at org.apache.commons.dbcp2.BasicDataSource.createDataSource(BasicDataSource.java:1903)
 at org.apache.commons.dbcp2.BasicDataSource.getConnection(BasicDataSource.java:1413)
 at i5.las2peer.services.gamificationGameService.database.DatabaseManager.getConnection(DatabaseManager.java:31)
 at i5.las2peer.services.gamificationGameService.GamificationGameService.getSeparateGameInfo(GamificationGameService.java:502)
 at jdk.internal.reflect.GeneratedMethodAccessor13.invoke(Unknown Source)
 at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.base/java.lang.reflect.Method.invoke(Method.java:568)
 at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
 at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
 at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
 at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
 at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
 at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:475)
 at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:397)
 at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
 at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255)
 at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
 at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
 at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
 at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
 at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
 at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
 at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
 at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684)
 at i5.las2peer.restMapper.RESTService.handle(RESTService.java:148)
 at jdk.internal.reflect.GeneratedMethodAccessor12.invoke(Unknown Source)
 at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.base/java.lang.reflect.Method.invoke(Method.java:568)
 at i5.las2peer.execution.ServiceHelper.execute(ServiceHelper.java:139)
 at i5.las2peer.security.ServiceAgentImpl.invoke(ServiceAgentImpl.java:453)
 at i5.las2peer.security.ServiceAgentImpl.lambda$handle$0(ServiceAgentImpl.java:479)
 at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
 at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
 at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
 at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: org.postgresql.util.PSQLException: FATAL: sorry, too many clients already
 at org.postgresql.core.v3.ConnectionFactoryImpl.doAuthentication(ConnectionFactoryImpl.java:623)
 at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:163)
 at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:215)
 at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:51)
 at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:225)
 at org.postgresql.Driver.makeConnection(Driver.java:466)
 at org.postgresql.Driver.connect(Driver.java:265)
 at org.apache.commons.dbcp2.DriverConnectionFactory.createConnection(DriverConnectionFactory.java:39)
 at org.apache.commons.dbcp2.PoolableConnectionFactory.makeObject(PoolableConnectionFactory.java:205)
 at org.apache.commons.dbcp2.BasicDataSource.validateConnectionFactory(BasicDataSource.java:2162)
 at org.apache.commons.dbcp2.BasicDataSource.createPoolableConnectionFactory(BasicDataSource.java:2148)
 ... 35 more
Request failed: java.lang.NullPointerException: Cannot invoke "java.sql.Connection.close()" because "conn" is null
java.lang.NullPointerException: Cannot invoke "java.sql.Connection.close()" because "conn" is null
 at i5.las2peer.services.gamificationGameService.GamificationGameService.getSeparateGameInfo(GamificationGameService.java:518)
 at jdk.internal.reflect.GeneratedMethodAccessor13.invoke(Unknown Source)
 at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.base/java.lang.reflect.Method.invoke(Method.java:568)
 at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
 at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
 at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
 at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
 at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
 at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:475)
 at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:397)
 at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
 at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255)
 at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
 at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
 at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
 at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
 at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
 at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
 at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
 at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684)
 at i5.las2peer.restMapper.RESTService.handle(RESTService.java:148)
 at jdk.internal.reflect.GeneratedMethodAccessor12.invoke(Unknown Source)
 at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 at java.base/java.lang.reflect.Method.invoke(Method.java:568)
 at i5.las2peer.execution.ServiceHelper.execute(ServiceHelper.java:139)
 at i5.las2peer.security.ServiceAgentImpl.invoke(ServiceAgentImpl.java:453)
 at i5.las2peer.security.ServiceAgentImpl.lambda$handle$0(ServiceAgentImpl.java:479)
 at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
 at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
 at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
 at java.base/java.lang.Thread.run(Thread.java:831)
```

**Comment by fxjordan** [comment]
Reproduced on localhost (default Postgres connection limit seems to be 100):

1. Run http://localhost:8080/gamification/games/list/separated about 100 times (200 OK, success response)
2. After 100 requests: 500 Internal Server error (and exception as above in console)

**Comment by fxjordan** [comment]
NullPointerException when calling `Connection.close()` i already fixed, but not the main issue.
Also, I tried closing the `PreparedStatment` instances after execution. However, this does not resolve the problem.

**Comment by fxjordan** [comment]
The following code in the `DatabaseManager` might be part of the problem. The data source is a **static** field, but is always re-initialized when the constructor is called (this is really bad code...). The constructor is called once per request, (!) Maybe turning the `DatabaseManager` into an actual singleton that manages the connection is a good idea.

https://github.com/rwth-acis/Gamification-Framework/blob/2f9572744da160af55277f3aa8e169ea162b5c58/GamificationGameService/src/main/java/i5/las2peer/services/gamificationGameService/database/DatabaseManager.java#L14-L32

**Comment by fxjordan** [comment]
Fixed and deployed as version 0.1.6
